src/hooks/useSignChallenge.ts

```diff
@@ -0,0 +1,123 @@
+import { useEffect, useState, useCallback, useRef } from 'react';
+import { useSignMessage } from 'wagmi';
+import { SIGNING_SERVICE_URL } from '../constants/constants';
+import { storageKeys } from '../constants/localStorage';
+import { SiweMessage } from 'siwe';
+
+export interface SiweSignatureData {
+  nonce: string;
+  signature: string;
+  expirationDate: string;
+}
+
+export function useSiweSignature(address?: `0x${string}`) {
+  const { signMessageAsync } = useSignMessage();
+  const [requiresSign, setRequiresSign] = useState(false);
+
+  // Used to wait for the modal interaction and/or return of the
+  // signing promise.
+  const signPromiseRef = useRef<{
+    resolve: (data: SiweSignatureData) => void;
+    reject: (reason: Error) => void;
+  } | null>(null);
+
+  const storageKey = `${storageKeys.SIWE_SIGNATURE_KEY_PREFIX}${address}`;
+
+  const checkStoredSignature = useCallback((): SiweSignatureData | null => {
+    if (!address) return null;
+
+    try {
+      const stored = localStorage.getItem(storageKey);
+      if (!stored) return null;
+
+      const data: SiweSignatureData = JSON.parse(stored);
+      return new Date(data.expirationDate) > new Date() ? data : null;
+    } catch {
+      localStorage.removeItem(storageKey);
+      return null;
+    }
+  }, [address, storageKey]);
+
```

Comment:
We want to return a promise that will resolve when the signing finishes (or fails). This is so we can use `checkAndWaitForSignature` and it's force counterpart on the `sep-10` process function.